src/main/java/org/radarcns/biovotion/BiovotionDeviceManager.java

```diff
@@ -197,19 +201,55 @@ public void close() throws IOException {
         if (isClosed()) {
             return;
         }
-        super.close();
         logger.info("Biovotion VSM Closing device {}", this);
         executor.shutdown();
 
+        super.close();
+    }
+
+    public void disconnect() {
```

Comment:
I suppose the disconnect() call is now idempotent in a single thread, but if disconnect is called from another thread, this will cause undefined behaviour. Please add a comment that it should only be called from a single thread (and ensure that this is the case), or synchronise the null-check + setting to null (using `AtomicReference` or a synchronised block with temporary local variable, or synchronise the entire disconnect method.

Reply:
Yes you are right, will def take a look at it in my next PR.